Fixes for the rewritefrom functionality

keithbare2 opened the pull request:
Closes #12

Gmail has been rejecting some of the messages sent to ezmlm-idx mailing lists I help run. Eventually I determined it was due to the outgoing messages having two different Cc headers: one from the original poster and then a second with the poster's address, the latter having been added by the rewritefrom logic (we use replytolist). I saw that issue #12 mentions the same problem happening with Reply-To headers.

We had not noticed the issue with encoded words that is perhaps the main complaint in issue #12.  (We tend to all be Americans with 7-bit ASCII names.)

In any case, the duplicated headers problem is causing us a lot of trouble. Essentially we lose messages whenever somebody tries to use reply-all to messages sent on our lists. So I've implemented a fix. And since it was in the same GitHub issue, I also took a swing at maintaining encoded words in the author name of the rewritten From header.

keithbare2 committed:
Fix comments in lib/author.c

The comments for author_name() and author_addr() did not accurately describe
what the code is actually doing.

keithbare2 committed:
Adjust construction of rewritten From lines

Copy rfc2047 encoded words in the original From line to the rewritten From
line without decoding them.

The encoded words also need to remain unquoted in the rewritten From line.
To accomplish this, do not quote the entire result of the AUTHOR_VIA_LIST
template substitution. Instead, if the original From line's author name text
was from a quoted string, comment or bare address, quote the author name
text before substituting. Otherwise, the original From line's author name
text (which will not contain a quoted string) is substituted verbatim.

keithbare2 committed:
Correct the ezmlm-test manual page

Running ezmlm-test does not require a working qmail installation.

Do not explicitly warn about the dot.qmail-__tstlist* files, because
they're created in __TSTDIR.

keithbare2 committed:
Avoid generating duplicate headers in ezmlm-send

When processing a message's headers:

 - Make note if Reply-To was seen.
 - Save off and do not copy the Cc header to the output immediately.
 - Save off and do not copy the From header to the output immediately.

At the end of the message headers, call rewrite_from().

When flagrewritefrom is set, either by the list's configuration or
because a DMARC reject policy applies to the message sender's address:

 - The From header is rewritten as before.
 - If replytolist is not enabled, a Reply-To header is only added
   if the original message did not have one. Rationale is that the
   original Reply-To header indicated where the post author is
   interested in receiving responses, not the From header.
 - When replytolist is enabled and the original message had a Cc header,
   the From address is appended to the existing Cc header.
 - When replytolist is enabled and the original message did not have a
   Cc header, generate a new one using the From address.

After rewrite_from() performs any of the applicable manipulations
described above, the resulting From and Cc headers are copied to the
output.

Update tests/551-ezmlm-send-rewritefrom:

 - Confirm Reply-To and Cc headers are not duplicated.
 - Validate behavior when replytolist is configured.

keithbare2 committed:
Add rewritefrom test with multipart message

keithbare2 commented:
The primary purpose of the update was to resolve an issue in my original changes that arose for MIME messages with a multipart Content-Type.  In that case, `flagbadpart` would get set on the header -> body transition, and inhibit output of the `line` containing the rewritten From and new Reply-To/Cc headers.  To resolve, the actual `line` content is output when `flagbadpart` is set, rather than assuming only a "\n" is required.

raboof commented:
This looks great, thanks so much for working on this! We're also running into the duplicate `Reply-To`/`Cc` issue, and at first glance c179dfd looks like a reasonable solution.

Perhaps it should be its own PR, to make it more self-contained and likely to get merged?

raboof commented on bin/ezmlm-send.c:
```diff
@@ -335,21 +339,39 @@ static void rewrite_from()
     stralloc_catb(&line,">\n",2);
-    stralloc_cats(&line,flagreplytolist ? "Cc:" : "Reply-To:");
+    if (flagreplytolist) {
+      if (cc.s) {
+        --cc.len;	/* remove '\n' */
+        stralloc_catb(&cc,",\n",2);
+      }
+      stralloc_catb(&cc,from.s,from.len);
```
This assumes `from` starts with whitespace (otherwise the appended value will not be considered part of the Cc header, but as the start of a new header), which is not guaranteed. Probably just insert a ` `?

keithbare2 replied:
I think I remember reading in RFC 5322 (or potentially one of the earlier RFCs defining the email message format) that whitespace is actually required between the `:` and header value.  So there should be some whitespace there from when the `From` header value was saved (lines 656-657):
```c
        } else if (case_startb(line.s,line.len,"From:")) {
	  stralloc_copyb(&from,line.s+5,line.len-5);
```

Though I recognize ezmlm-send could be fed a non-compliant message—so it's probably not a bad idea to ensure there's leading whitespace for the continuation line, as you suggest, especially if ezmlm-send isn't otherwise assuming there is whitespace separating the header names and values.

raboof replied:
it's definitely common but I didn't find it in https://www.rfc-editor.org/rfc/rfc5322#section-3, and the tests didn't complain when I created a `From:` that was not followed by whitespace.

keithbare2 replied:
Addressed by commit [17d9159](https://github.com/bruceg/ezmlm-idx/pull/20/commits/17d91594efebf6bff5abc35680257ab359ecda6e).

keithbare2 committed:
Handle From header lacking whitespace after colon

When rewritefrom needs to append the original From address to an existing Cc
header the from address is placed on the next line. Whitespace is required
at the beginning of the line so the added line is treated as an extension of
the previous header. Ensure there is whitespace at the beginning of the
line, in case there was not any after the colon in the From header.

keithbare2 commented:
> This looks great, thanks so much for working on this! We're also running into the duplicate `Reply-To`/`Cc` issue

You're welcome.  Starting in March 2023, I've been running several [cmucc](https://github.com/cmucc) mailing lists with the changes I posted here and haven't noticed any problems.

In preparation for Google's sender requirements going into effect, I wanted to switch everything to use rewritefrom unconditionally.  I found myself rebuilding ezmlm-idx and was reminded that I faced a few problems building and installing code based on the master branch here.  I collected my fixes for those problems and posted them in #22.  I figure that might help if you or anybody else wants to build and run with changes from this PR.

> \[A]t first glance [c179dfd](https://github.com/bruceg/ezmlm-idx/commit/c179dfdb0a307402fe0ac7a87a0301e6683c0b93) looks like a reasonable solution.

> Perhaps it should be its own PR, to make it more self-contained and likely to get merged?

The only problem with using that commit on its own is that the tests I wrote have a dependency on the way [3bf20d9](https://github.com/bruceg/ezmlm-idx/pull/20/commits/3bf20d94139da57bc370ae2e2dba1491ad92b067) changed the quoting of the From header's display name.  Should be straightforward to resolve though.

I haven't heard anything from the owner or anybody else that could merge a PR.  I certainly could create a new PR with just the duplicate header fix if they ask.
